### PR TITLE
Add source code Chapter 21

### PR DESCRIPTION
**Pull request by Peter-VanWeert**
Solutions to the remaining exercises will follow later...

### CODE REVIEW

File: Examples/NoModules/Chapter 21/Ex21_01/Ex21_01.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#include <algorithm>
+
+class MyClass { /* just a dummy class */ };
+
+int main()
+{
+  std::vector<int&> v;
+
+  MyClass one, other;
+  auto biggest = std::max(one, other);
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 21/Ex21_01/Ex21_01.cpp:14]: uniform init

File: Examples/NoModules/Chapter 21/Ex21_03/Ex21_03.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Violating constraints of uninstantiated class members
+#include "Array.h"
+#include <memory>  // For std::unique_ptr<>
+
+// Assert that Array<std::unique_ptr<int>> is a valid type
+static_assert(requires { typename Array<std::unique_ptr<int>>; });
+
+int main()
+{
+  Array<std::unique_ptr<int>> tenSmartPointers(10);
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 21/Ex21_03/Ex21_03.cpp:10]: uniform init

File: Examples/NoModules/Chapter 21/Ex21_03/Ex21_03.cpp [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Violating constraints of uninstantiated class members
+#include "Array.h"
+#include <memory>  // For std::unique_ptr<>
+
+// Assert that Array<std::unique_ptr<int>> is a valid type
+static_assert(requires { typename Array<std::unique_ptr<int>>; });
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 21/Ex21_03/Ex21_03.cpp:6]: doesn't compile on VC++. Bug in VC++ or bug in code?
```
1>Ex21_03.cpp(6,15): error C2059: syntax error: 'requires'
1>Ex21_03.cpp(6,24): error C2143: syntax error: missing ';' before '{'
1>Ex21_03.cpp(6,24): error C2143: syntax error: missing ')' before ';'
1>Ex21_03.cpp(6,24): error C2447: '{': missing function header (old-style formal list?)
1>Ex21_03.cpp(6,65): error C2059: syntax error: ')'
```

File: Examples/NoModules/Chapter 21/Ex21_04/Ex21_04.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+{
+  std::cout << "Distance determined in constant time: ";
+  return last - first;
+}
+
+int main()
+{
+  std::list l{ 'a', 'b', 'c' };
+  std::vector v{ 1, 2, 3, 4, 5 };
+  float a[] = { 1.2f, 3.4f, 4.5f };
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 21/Ex21_04/Ex21_04.cpp:29]: uniform init

File: Exercises/NoModules/Chapter 21/Soln21_02/Soln21_02.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+int main()
+{
+  std::vector<Box> boxes{ {2.0, 2.0, 3.0}, {1.0, 3.0, 2.0}, {1.0, 2.0, 1.0}, {2.0, 3.0, 3.0} };
+
+  // Step 1: create a view where begin() and end() are different.
+  // The result of the take_while() range adapter has this property
+  // (or, at least, it should have as per its specification in the Standard).
+  auto first_boxes = boxes | take_while([](const Box& box) { return box.volume() < 15; });
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 21/Soln21_02/Soln21_02.cpp:78]: uniform init
same on a few other places below.

File: Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
+// Exercise 21-3 
+#include <iostream>
+#include <concepts>
+#include <span>
+#include <array>
+#include <string>
+
+/* Not requested, but just for fun, here is a concept that prescrives
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp:8]: prescribes

File: Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
+// Exercise 21-3 
+#include <iostream>
+#include <concepts>
+#include <span>
+#include <array>
+#include <string>
+
+/* Not requested, but just for fun, here is a concept that prescrives 
+   a complate set of operations one would require to take an average */
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp:9]: complete

File: Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+int main()
+{
+  std::array values_odd{ 1, 2, 3, 4, 5, 6, 7 };
+  std::cout << medianOfSorted(std::span{ values_odd }) << std::endl;
+
+  std::array values_even{ 1., 2., 3., 4., 5., 6. };
+  std::cout << medianOfSorted(std::span{ values_even }) << std::endl;
+
+  std::string strings_odd[] = { "1", "2", "3", "4", "5" };
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp:46]: uniform init
also below

File: Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+template <typename T, size_t N>
+auto& medianOfSorted(std::span<T, N> span)
+{
+  return span[N / 2];
+}
+
+template <Averagable T, size_t N> requires (N % 2 == 0)
+auto medianOfSorted(std::span<T, N> span)
+{
+  return (span[N / 2] + span[N / 2 + 1]) / 2;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 21/Soln21_03/Soln21_03.cpp:35]: this is not correct. with this, the median of the following
`std::array values_even{ 1., 2., 3., 4., 5., 6. };`
is 4.5 while it should be 3.5.
So:
`return (span[N / 2 - 1] + span[N / 2]) / 2;`
?
But then you need to be careful that N/2-1 is a valid index.

**Review by MarcGregoire — CHANGES_REQUESTED** [review]

File: Examples/NoModules/Chapter 21/Ex21_03/Ex21_03.cpp (hunk 3 above)
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 21/Ex21_03/Ex21_03.cpp:6]: Compiles fine with GCC, so I'm thinking not implemented yet in VC++?

**Pull request merged**